File: packages/vscode-extension/src/activity-log.js
/**
 * In-memory activity log for the IDE bridge — a small ring buffer of events
 * (tool calls, connects) that the status bar, sidebar tree, and webview
 * dashboard all render from. Pure (no `vscode`), so it is unit-testable; the UI
 * just subscribes via onChange.
 */
class ActivityLog {
  constructor({ max = 200 } = {}) {
    this._max = max;
    this._entries = [];
    this._listeners = new Set();
    this._counts = { tool: 0, connect: 0, error: 0 };
  }

  /**
   * Append an event. Shape: { type:"tool"|"connect", tool?, ok?, error?,
   * argsSummary?, ts }. Returns the stored (normalized) entry.
   */
  record(entry) {
    const e = { ts: Date.now(), ...entry };
    this._entries.push(e);
    if (this._entries.length > this._max) this._entries.shift();
    if (e.type === "tool") this._counts.tool++;
    if (e.type === "connect") this._counts.connect++;
    if (e.ok === false) this._counts.error++;
    for (const l of this._listeners) {
      try {
        l(e);
      } catch {
        /* a bad listener must not break logging */
      }
    }
    return e;
  }

  /** All entries, oldest first. */
  entries() {
    return this._entries.slice();
  }

  /** The most recent `n` entries, newest first. */
  recent(n = 20) {
    return this._entries.slice(-n).reverse();
  }

  counts() {
    return { ...this._counts };
  }

  clear() {
    this._entries = [];
    for (const l of this._listeners) {
      try {
        l(null);
      } catch {
        /* ignore */
      }
    }
  }

  /** Subscribe; returns an unsubscribe fn. */
  onChange(cb) {
    this._listeners.add(cb);
    return () => this._listeners.delete(cb);
  }
}

/** Short, non-sensitive summary of a tool call's arguments for display. */
function summarizeArgs(toolName, args) {
  if (!args || typeof args !== "object") return "";
  if (toolName === "openDiff") {
    const p = typeof args.path === "string" ? args.path : "";
    return p ? shortenPath(p) : "";
  }
  if (toolName === "getDiagnostics" && typeof args.path === "string") {
    return shortenPath(args.path);
  }
  return "";
}

function shortenPath(p) {
  const parts = String(p).replace(/\\/g, "/").split("/").filter(Boolean);
  return parts.length <= 2 ? p : "…/" + parts.slice(-2).join("/");
}

module.exports = { ActivityLog, summarizeArgs, shortenPath };

File: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/ActivityLog.java
package com.chainlesschain.ide;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/**
 * In-memory activity log for the IDE bridge — a bounded ring buffer of events
 * (tool calls) that the "Show Activity" dialog + status renders from. The Java
 * twin of the VS Code extension's activity-log.js (which feeds its webview
 * dashboard). Pure (no IntelliJ SDK), so it's unit-testable; McpServer records
 * into it via an {@link com.chainlesschain.ide.McpServer.ActivityListener} and
 * the glue renders {@link #recent}/{@link #counts}.
 *
 * Thread-safe: McpServer serves tool calls on a pooled thread while the EDT
 * reads for the dialog, so record/read are synchronized.
 */
public final class ActivityLog {

    /** One recorded event. */
    public static final class Entry {
        public final long ts;
        public final String type;    // "tool" | "connect"
        public final String tool;    // nullable (connect has none)
        public final boolean ok;
        public final String summary; // short arg summary, nullable

        public Entry(long ts, String type, String tool, boolean ok, String summary) {
            this.ts = ts;
            this.type = type;
            this.tool = tool;
            this.ok = ok;
            this.summary = summary;
        }
    }

    /** Rolling totals. */
    public static final class Counts {
        public final int tool;
        public final int connect;
        public final int error;

        Counts(int tool, int connect, int error) {
            this.tool = tool;
            this.connect = connect;
            this.error = error;
        }
    }

    private final int max;
    private final Deque<Entry> entries = new ArrayDeque<Entry>();
    private int toolCount, connectCount, errorCount;

    public ActivityLog(int max) {
        this.max = max <= 0 ? 200 : max;
    }

    /** Append an event; drops the oldest past the cap. Returns the stored entry. */
    public synchronized Entry record(long ts, String type, String tool, boolean ok, String summary) {
        Entry e = new Entry(ts, type, tool, ok, summary);
        entries.addLast(e);
        while (entries.size() > max) entries.pollFirst();
        if ("tool".equals(type)) toolCount++;
        if ("connect".equals(type)) connectCount++;
        if (!ok) errorCount++;
        return e;
    }

    /** The most recent {@code n} entries, newest first. */
    public synchronized List<Entry> recent(int n) {
        List<Entry> out = new ArrayList<Entry>();
        java.util.Iterator<Entry> it = entries.descendingIterator();
        while (it.hasNext() && out.size() < n) out.add(it.next());
        return out;
    }

    public synchronized Counts counts() {
        return new Counts(toolCount, connectCount, errorCount);
    }

    public synchronized int size() {
        return entries.size();
    }

    /**
     * A plain-text report the "Show Activity" dialog shows: a header line with
     * the running totals, then the most recent {@code n} calls newest-first
     * (HH:mm:ss · ✓/✗ tool · summary). Timestamps use the supplied formatter so
     * the caller controls the zone (and tests stay deterministic).
     */
    public synchronized String formatReport(int bridgePort, int recentN, TimeFmt fmt) {
        Counts c = counts();
        StringBuilder sb = new StringBuilder("ChainlessChain IDE bridge — activity\n\n");
        sb.append(bridgePort > 0
                ? "bridge on 127.0.0.1:" + bridgePort + " (server \"ide\")\n"
                : "bridge stopped\n");
        sb.append("tool calls: ").append(c.tool)
          .append("   ·   connections: ").append(c.connect)
          .append("   ·   errors: ").append(c.error).append("\n\n");
        List<Entry> recent = recent(recentN);
        if (recent.isEmpty()) {
            sb.append("(no tool calls yet — the agent hasn't used the IDE bridge)\n");
            return sb.toString();
        }
        sb.append("recent (newest first):\n");
        for (Entry e : recent) {
            String name = "tool".equals(e.type) ? (e.tool == null ? "?" : e.tool) : "connection";
            sb.append("  ").append(fmt.format(e.ts))
              .append("  ").append(e.ok ? "✓" : "✗")
              .append(" ").append(name);
            if (e.summary != null && !e.summary.isEmpty()) sb.append("  ").append(e.summary);
            sb.append("\n");
        }
        return sb.toString();
    }

    /** Time formatter seam (the glue passes a local HH:mm:ss; tests a stub). */
    public interface TimeFmt {
        String format(long ts);
    }

    /** Short, non-sensitive summary of a tool call's args (mirrors VS summarizeArgs). */
    public static String summarizeArgs(String toolName, java.util.Map<String, Object> args) {
        if (args == null) return "";
        Object path = args.get("path");
        if (("openDiff".equals(toolName) || "getDiagnostics".equals(toolName))
                && path instanceof String) {
            return shortenPath((String) path);
        }
        return "";
    }

    static String shortenPath(String p) {
        if (p == null) return "";
        String[] parts = p.replace('\\', '/').split("/");
        List<String> nonEmpty = new ArrayList<String>();
        for (String s : parts) if (!s.isEmpty()) nonEmpty.add(s);
        if (nonEmpty.size() <= 2) return p;
        return "…/" + nonEmpty.get(nonEmpty.size() - 2) + "/" + nonEmpty.get(nonEmpty.size() - 1);
    }
}
